# Phlag

**Phlag** detects and flags phylogenetic anomalies across the genome.

Given a species tree and a sequence of gene trees, Phlag aims to detect strong deviations from the multi-species coalescent (MSC) process using quadripartition quartet scores (QQS) and a hidden Markov model (HMM).

The input is a sequence of gene trees, ideally sampled uniformly across a chromosome. The expected output is a set of "flagged" subsequences corresponding to model violations (e.g., admixture, recombination suppression, or dramatic changes in effective population size) and an updated species tree with new branch length (in coalescent units), estimated by ignoring the flagged gene trees.

## Installation

You will need a Python 3.9 environment. To install, follow the steps below:

```shell
micromamba create -n phlag python=3.9 -y
micromamba activate phlag
git clone https://github.com/bo1929/phlag.git
cd phlag
pip install -r requirements.txt
./build.sh
```

## Quickstart with a toy example

Suppose we have a species tree (`test/neoaves.nwk`) and an ordered set of gene trees (`test/emission.gtrees`).
This gene tree sequence is a mixture of two coalescent with recombination processes, with one key difference: the effective population size for certain *Charadriiformes* increased 10-fold.

The mixture happens in a way that only a specific loci are affected by this change.
In this particular example, 150 out of 1500 consecutive gene trees are generated by the alternative process.
Our goal is to detect this region (gene trees in indices [913, 1062]) using Phlag.


### Input
* **Species tree**: A Newick species tree with labeled internal nodes.
* **Gene trees**: One Newick tree per line, ordered along the genome (not necessarily with labeled internal nodes).
* **Focal branch(es)**: The label of the clade(s) to be targeted. In this example, the clade under suspicion is labeled `N159`.

### Usage
```shell
python phlag.py \
  --species-tree test/neoaves.nwk \
  --gene-trees  test/emission.gtrees \
  -c N159 \
  --num-ters 10 \
  --read-qqs-freqs test/qqs.tsv \
  --output-file results-neoaves-N159.txt
```
The quadripartition quartet scores are already computed and stored in `test/qqs.tsv`.
You can directly read those to avoid extra computation for this toy example.
If you do not give `--read-qqs-freqs`, QQS values need to be computed, which adds to the running time.
Note that you can also save QQS values to read later for different runs (i.e., changing the focal branch or the hyperparameters) using `write-qqs-freqs`.

For other options, run `python phlag.py --help`.

**Note**: If your data lacks strong deviations from the MSC, you may see few or no flagged trees.
Conversely, an incorrect/unreliable species tree may result in too many flagged gene trees.
In such cases, you can attempt to adjust the hyperparameters to detect anomalies at the resolution you desire.

### Output
The output is rather simple to interpret:
- **Header (Lines 1-5)**: Auxiliary information prefixed with `#`:
    * Simply the invoked command and the version information.
    * The initial species tree with branch lengths in CU, estimated using the entire gene tree sequence.
    * The same species tree topology, but branch lengths are estimated only based on the gene trees that are not flagged.
    * The labels of the clades for the focal branches that are used, and their final branch lengths.
    * Distances between the emission distributions of the two states, separately for all the focal branches following the order given in the above line.
- **State label**: `1` corresponds to flagged gene trees; `0` corresponds to MSC-compliant gene trees.
- **Posterior probabilities**: The decoding probability for each gene tree.
